### PR TITLE
[FEAT] Implementation of Xception NN

### PR DESCRIPTION
**Pull request by prkhrsrvstv1**
**Closes**: #19 

* Added implementation for Xception
* Added relevant diagram from the paper

**Ref.**: Chollet, François. “Xception: Deep Learning with Depthwise Separable Convolutions.” _2017 IEEE Conference on Computer Vision and Pattern Recognition (CVPR)_ (2017): 1800-1807.

### CODE REVIEW

**Comment by EverLookNeverSee** [comment]
@prkhrsrvstv1 Thank you for this PR, i will review it as soon as possible.

File: models/xception/xception.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from tensorflow.keras.layers import Activation, Add, BatchNormalization, Conv2D, Dense, GlobalAveragePooling2D, Input, MaxPool2D, SeparableConv2D
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:1]: Add an initial docstring on top of the script
Use parenthesis in import statement in order to observe line length

File: models/xception/xception.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+from tensorflow.keras.models import Model
+
+# 299 x 299 x 3 images
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:4]: Add a meaningful comment e.g. Defining the input layer

File: models/xception/xception.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+for filters in [128, 256, 728]:
+    res = Conv2D(filters=filters, kernel_size=(1, 1), strides=(2, 2), padding="same", use_bias=False)(outputs)
+    res = BatchNormalization()(res)
+    for _ in range(2):
+        outputs = Activation("relu")(outputs)
+        outputs = SeparableConv2D(filters=filters, kernel_size=(3, 3), padding="same", use_bias=False)(outputs)
+        outputs = BatchNormalization()(outputs)
+    outputs = MaxPool2D(pool_size=(3, 3), strides=(2, 2), padding="same")(outputs)
+    outputs = Add()([res, outputs])
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:24]: Define entry flow section as a function

File: models/xception/xception.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+# Entry flow
+outputs = Conv2D(filters=32, kernel_size=(3, 3), strides=(2, 2), use_bias=False)(inputs)
+outputs = BatchNormalization()(outputs)
+outputs = Activation("relu")(outputs)
+
+outputs = Conv2D(filters=64, kernel_size=(3, 3), use_bias=False)(outputs)
+outputs = BatchNormalization()(outputs)
+outputs = Activation("relu")(outputs)
+
+for filters in [128, 256, 728]:
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:16]: for filter in filters

File: models/xception/xception.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+# Middle flow
+for _ in range(8):
+    res = outputs
+    for _ in range(3):
+        outputs = Activation("relu")(outputs)
+        outputs = SeparableConv2D(filters=728, kernel_size=(3, 3), padding="same", use_bias=False)(outputs)
+        outputs = BatchNormalization()(outputs)
+    outputs = Add()([res, outputs])
+
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:34]: Define middle flow as a function

File: models/xception/xception.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
+from tensorflow.keras.layers import Activation, Add, BatchNormalization, Conv2D, Dense, GlobalAveragePooling2D, Input, MaxPool2D, SeparableConv2D
+from tensorflow.keras.models import Model
+
+# 299 x 299 x 3 images
+inputs = Input(shape=(299, 299, 3))
+
+# Entry flow
+outputs = Conv2D(filters=32, kernel_size=(3, 3), strides=(2, 2), use_bias=False)(inputs)
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:8]: Use meaningful and distinguishable variable names

File: models/xception/xception.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    outputs = Activation("relu")(outputs)
+    outputs = SeparableConv2D(filters=filters, kernel_size=(3, 3), padding="same", use_bias=False)(outputs)
+    outputs = BatchNormalization()(outputs)
+outputs = MaxPool2D(pool_size=(3, 3), strides=(2, 2), padding="same")(outputs)
+outputs = Add()([res, outputs])
+
+for filters in [1536, 2048]:
+    outputs = SeparableConv2D(filters=filters, kernel_size=(3, 3), padding="same", use_bias=False)(outputs)
+    outputs = BatchNormalization()(outputs)
+    outputs = Activation("relu")(outputs)
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:48]: Define exit flow as a function

File: models/xception/xception.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+outputs = GlobalAveragePooling2D()(outputs)
+
+# Optional fully-connected layer(s)
+outputs = Dense(units=2048)(outputs)
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:53]: Use meaningful and distinguishable variable names

File: models/xception/README.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Xception
+
```
Comment (EverLookNeverSee) [inline models/xception/README.md:2]: Also add xception model to main README file

**Review by EverLookNeverSee — CHANGES_REQUESTED** [review]

File: models/xception/xception.py (hunk 4 above)
Comment (prkhrsrvstv1) [reply models/xception/xception.py]: Renamed this variable to `num_filters` for better clarity.

File: README.md [hunk 10]
```diff
@@ -28,7 +28,4 @@
 Neural Networks
 * [VGG-16](models/vgg_16/vgg-16.py) - Very Deep Convolutional Networks For Large Scale Image Recognition
 * [Inception-v1](models/inception_v1/inception-v1.py) - Going Deeper With Convolutions
-
-
-## License
-This project licensed under the MIT License - see the [LICENSE](LICENSE) file for more details.
+* [Xception](models/xception/xception.py) - Deep Learning with Depthwise Separable Convolutions
```
Comment (EverLookNeverSee) [inline README.md:31]: LGTM!

File: models/xception/xception.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    Author: Prakhar Srivastava - @prkhrsrvstv1
+"""
+from typing import Any
+
+from tensorflow.keras.layers import (Activation, Add, BatchNormalization, Conv2D, Dense, GlobalAveragePooling2D, Input,
+                                     MaxPool2D, SeparableConv2D)
+from tensorflow.keras.models import Model
+
+
+def convolutional_unit(conv_inputs, num_filters, kernel_size, strides=(1, 1), pre_activation=False,
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:13]: Add function docstring and use type hint for parameters

File: models/xception/xception.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        conv_inputs = Activation("relu")(conv_inputs)
+    conv_outputs = conv_layer(filters=num_filters, kernel_size=kernel_size, strides=strides, padding="same",
+                              use_bias=False)(conv_inputs)
+    conv_outputs = BatchNormalization()(conv_outputs)
+    if post_activation:
+        conv_outputs = Activation("relu")(conv_outputs)
+    return conv_outputs
+
+
+def separable_convolutional_unit(sep_conv_inputs, num_filters, pre_activation=True, post_activation=False):
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:25]: Add function docstring and use type hint for parameters

File: models/xception/xception.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    return convolutional_unit(sep_conv_inputs, num_filters, (3, 3), pre_activation=pre_activation,
+                              post_activation=post_activation, conv_layer=SeparableConv2D)
+
+
+def entry_flow(entry_inputs):
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:30]: function docstring and type hints

File: models/xception/xception.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    for num_filters in [128, 256, 728]:
+        res = convolutional_unit(entry_outputs, num_filters, (1, 1), (2, 2), post_activation=False)
+        for _ in range(2):
+            entry_outputs = separable_convolutional_unit(entry_outputs, num_filters)
+        entry_outputs = MaxPool2D(pool_size=(3, 3), strides=(2, 2), padding="same")(entry_outputs)
+        entry_outputs = Add()([res, entry_outputs])
+    return entry_outputs
+
+
+def middle_flow(middle_inputs):
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:45]: Add function docstring

File: models/xception/xception.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    middle_outputs = middle_inputs
+    for _ in range(8):
+        res = middle_outputs
+        for _ in range(3):
+            middle_outputs = separable_convolutional_unit(middle_outputs, 728)
+        middle_outputs = Add()([res, middle_outputs])
+    return middle_outputs
+
+
+def exit_flow(exit_inputs):
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:56]: Add function docstring

**Review by EverLookNeverSee — CHANGES_REQUESTED** [review]

**Comment by EverLookNeverSee** [comment]
Pull new changes to your branch in order to prevent conflicts

**Comment by prkhrsrvstv1** [comment]
@EverLookNeverSee gentle reminder for this PR.

File: README.md [hunk 16]
```diff
@@ -28,6 +28,7 @@
 Neural Networks
 * [VGG-16](models/vgg_16/vgg-16.py) - Very Deep Convolutional Networks For Large Scale Image Recognition
 * [Inception-v1](models/inception_v1/inception-v1.py) - Going Deeper With Convolutions
+* [Xception](models/xception/xception.py) - Deep Learning with Depthwise Separable Convolutions
```
Comment (EverLookNeverSee) [inline README.md:31]: LGTM

File: models/xception/xception.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+        conv_inputs: Input tensor
+        num_filters: Number of filters to use in the convolutional layer
+        kernel_size: Kernel size for the convolutional layer
+        strides: Strides size for the convolutional layer
+        pre_activation: Whether or not to perform activation before the convolutional layer. False by default.
+        post_activation: Whether or not to perform activation after the batch-normalization layer. True by default.
+        conv_layer: The type of Convolutional layer to use. Should be either Conv2D or SeparableConv2D
+
+    Returns:
+        A 4+D Tensor obtained after passing through activation, convolutional, and batch-normalization layers
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:38]: LGTM - I will change the function docstring style later on.

File: models/xception/xception.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+    Args:
+        sep_conv_inputs: Input tensor
+        num_filters: Number of filters to use in the convolutional layer
+        pre_activation: Whether or not to perform activation before the convolutional layer. True by default.
+        post_activation: Whether or not to perform activation after the batch-normalization layer. False by default.
+
+    Returns:
+        A 4+D Tensor obtained after passing through activation, convolutional, and batch-normalization layers
+    """
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:69]: LGTM - I will change the function docstring style later on.

Comment (EverLookNeverSee) [inline models/xception/xception.py:69]: LGTM - I will change the function docstring style later on.

File: models/xception/xception.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+def middle_flow(middle_inputs: Tensor) -> Tensor:
+    """Middle Flow
+
+    Implements the second of the three broad parts of the model
+
+    Args:
+        middle_inputs: Tensor output generate by the Entry Flow, having shape [*, new_rows, new_cols, 728]
+
+    Returns:
+        Output tensor of shape [*, new_rows, new_cols, 728]
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:109]: LGTM - I will change the function docstring style later on.

File: models/xception/xception.py [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+
+    Implements the second of the three broad parts of the model. Includes the optional fully-connected layers,
+    and the logistic regression segment of the model.
+
+    Args:
+        exit_inputs: Tensor output generated by the Middle Flow segment, having shape [*, new_rows, new_cols, 728]
+
+    Returns:
+        Output tensor of shape [*, 1000], representing the classifier output for 1000 classes
+
```
Comment (EverLookNeverSee) [inline models/xception/xception.py:133]: LGTM - I will change the function docstring style later on.

**Review by EverLookNeverSee — APPROVED** [review]

**Comment by EverLookNeverSee** [comment]
@prkhrsrvstv1 Pull new changes to your branch and push back in order to prevent conflicts

**Comment by prkhrsrvstv1** [comment]
@EverLookNeverSee Rebased the feature branch with upstream master. No new code changes since last approval.

**Comment by EverLookNeverSee** [comment]
You have cg_dev branch on your origin

**Comment by prkhrsrvstv1** [comment]
Merged `cg_dev` with upstream main. But `feat/implementation-of-xception-nn` already has those changes. Let me know if I should delete `cg_dev` from my fork.

**Comment by EverLookNeverSee** [comment]
@prkhrsrvstv1 Thank you for this great PR

**Comment by prkhrsrvstv1** [comment]
@EverLookNeverSee Could you by any chance add the `hacktoberfest` to the repo topics and `hacktoberfest-accepted` label to this PR?

I was trying to get 4 PRs merged and the contributions are not recognised without those conditions 😁

**Comment by EverLookNeverSee** [comment]
@prkhrsrvstv1 There is an issue in your code:
https://github.com/EverLookNeverSee/diag2model/blob/50241775a1331b7b1975c831e89845a4e89e371e/models/xception/xception.py#L43-L44
parameter rank unfilled
Please fix it.

**Comment by EverLookNeverSee** [comment]
parameter rank unfilled
@

**Comment by prkhrsrvstv1** [comment]
Sure, @EverLookNeverSee .

I'll create another PR shortly.
